src/storage-controller/src/lib.rs

```diff
@@ -3259,52 +3259,72 @@ where
     /// collections, they are not derived from a state at some time `t` and we
     /// cannot maintain a desired state for them.
     ///
-    /// Returns a map with latest unpacked row per id.
+    /// Returns a map with latest unpacked row per key, if the key is a single
+    /// `GlobalId`. Otherwise returns an empty map.
```

Comment:
This is annoying. For the source/sink status history (both keyed by ID), we need that latest value to give it to the `CollectionStatusManager`. The replica status history is keyed by (replica ID, process ID), but we also don't need anything returned for it, so things work out.

I'm not sure if we can have a nicer API. We might be able to return a `BTreeMap<Vec<Datum>, Row>>`, which would work with any kind of key. But returning datums seems difficult because they contain non-static lifetimes. Also the caller code for the source/sink status history would become more complicated because it'd have to unwrap the `GlobalId` from the `Vec<Datum>`.

Reply:
I like the idea of returning `BTreeMap<K, Row>` in all cases. I'd propose that the caller have to pass in a function that performs key extraction given a `Vec<Datum>`:

```rust
   async fn partially_truncate_status_history<F, K>(
        &mut self,
        collection: IntrospectionType,
        write_handle: &mut WriteHandle<SourceData, (), T, Diff>,
        extract_key: Fn(Vec<Datum>) -> Result<K, ()>,
    ) -> BTreeMap<K, Row> {
```

Reply:
I've decided to pass a `SourceHistoryDesc<K>` type instead that contains an `extract_key` function but also specifies other collection-specific details. This way, there is only a single place that needs to know about the schema of a collection and the knowledge is not split between `partially_truncate_status_history` and its caller.

Reply:
That's gorgeous!